### PR TITLE
Feature/hpa compatible

### PR DESCRIPTION
**Pull request by ringtail**
Add hpa compatible. Usage Case: 
```
apiVersion: autoscaling.alibabacloud.com/v1beta1
kind: CronHorizontalPodAutoscaler
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: cronhpa-sample
spec:
   scaleTargetRef:
      apiVersion: autoscaling/v1
      kind: HorizontalPodAutoscaler
      name: nginx-deployment-hpa
   jobs:
   - name: "scale-down"
     schedule: "30 */1 * * * *"
     targetSize: 1
   - name: "scale-up"
     schedule: "0 */1 * * * *"
     targetSize: 3
```
scenes:    

| hpa(min/max)  | cronhpa  | deployment |  result  |      
| --- | --- | --- | --- |      
| 1/10                |     5          |  5           | hpa(1/10) deployment 5|  
| 1/10              |  4            |      5             | hpa(1/10) deployment 5 |     
| 1/10               |    6          |      5             | hpa(6/10) deployment 6|  
| 5/10              |  5             | 5                   | hpa(5/10) deployment 5|    
| 5/10              | 4              | 5                   |hpa(4/10) deployment 5| 
| 5/10             | 6              | 5                   | hpa(6/10) deployment 6 | 
| 5/10             | 11             | 5                   | hpa(11/11) deployment 11 |

### CODE REVIEW

**Comment by ringtail** [comment]
Fix #13

File: pkg/controller/cronhorizontalpodautoscaler/cronjob.go [hunk 1]
```diff
@@ -86,60 +90,156 @@ func (ch *CronJobHPA) Ref() *TargetRef {
...
+
+	mappings, err := ch.mapper.RESTMappings(targetGK)
+	if err != nil {
+		return "", fmt.Errorf("Failed to create mapping,because of %s", err.Error())
+	}
+
+	found := false
+	for _, mapping := range mappings {
+		targetGR = mapping.Resource.GroupResource()
+		scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, ch.TargetRef.RefName)
```
Comment (yancl) [inline pkg/controller/cronhorizontalpodautoscaler/cronjob.go]: `scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, ch.TargetRef.RefName)` 
 should be
`scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, hpa.Spec.ScaleTargetRef.Name)` ?

File: pkg/controller/cronhorizontalpodautoscaler/cronjob.go [hunk 2]
```diff
@@ -86,60 +90,156 @@ func (ch *CronJobHPA) Ref() *TargetRef {
...
+	}
+
+	if updateHPA {
+		err = ch.client.Update(ctx, hpa)
 		if err != nil {
-			log.Warningf("Failed to scale (namespace: %s;kind: %s;name: %s) to %d,because of %s", ch.TargetRef.RefNamespace, ch.TargetRef.RefKind, ch.TargetRef.RefName, ch.DesiredSize, err.Error())
-		} else {
+			return "", err
+		}
+	}
```
Comment (yancl) [inline pkg/controller/cronhorizontalpodautoscaler/cronjob.go]:
```
        if updateHPA {
                err = ch.client.Update(ctx, hpa)
                if err != nil {
                        return "", err
                }
        }

        if hpa.Status.CurrentReplicas > ch.DesiredSize {
                // skip change replicas and exit
                return "", nil
        }
```

I think the *skip* should after the *updateHPA*, or else we won't have a chance to scale down the deployment after scale up the HPA. am i right? 
suppose the initial hpa is (min,desired,max) is (0,3,10), and changed to be (5,5,10) after cronhpa scale up. then scale down to 3 replicas by cronhpa. if we *skip* before *updateHPA*, then we will always be at least 5 replicas.

**Comment by NUCsimple** [comment]
Add hpa compatible. Usage Case:
```
apiVersion: autoscaling.alibabacloud.com/v1beta1
kind: CronHorizontalPodAutoscaler
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: cronhpa-sample
spec:
   excludeDates:
   - ' * * * 12 2 *'
   scaleTargetRef:
      apiVersion: autoscaling/v1
      kind: HorizontalPodAutoscaler
      name: nginx-deployment-hpa
   jobs:
   - name: "scale-down"
     schedule: "30 */1 * * * *"
     targetSize: 1
```
scenes:

hpa(min/max) | cronhpa | deployment | result
-- | -- | -- | --
5/10 | 1 | 5 | hpa(5/10) deployment 5
--------------
```
apiVersion: autoscaling.alibabacloud.com/v1beta1
kind: CronHorizontalPodAutoscaler
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: cronhpa-sample
spec:
   scaleTargetRef:
      apiVersion: autoscaling/v1
      kind: HorizontalPodAutoscaler
      name: nginx-deployment-hpa
   jobs:
   - name: "scale-down"
     schedule: "* * * 12 2 *"
     targetSize: 1
```
scenes:

hpa(min/max) | cronhpa | deployment | result
-- | -- | -- | --
5/10 | 1 | 1 | hpa(5/10) deployment 1

-----------
```
apiVersion: autoscaling.alibabacloud.com/v1beta1
kind: CronHorizontalPodAutoscaler
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: cronhpa-sample
spec:
   scaleTargetRef:
      apiVersion: autoscaling/v1
      kind: HorizontalPodAutoscaler
      name: nginx-deployment-hpa
   jobs:
   - name: "scale-down"
     schedule: "* * 11 12 2 *"
     targetSize: 1
```
scenes:

hpa(min/max) | cronhpa | deployment | result
-- | -- | -- | --
5/10 | 1 | 1 | hpa(5/10) deployment 1
--------
```
apiVersion: autoscaling.alibabacloud.com/v1beta1
kind: CronHorizontalPodAutoscaler
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: cronhpa-sample
spec:
   scaleTargetRef:
      apiVersion: autoscaling/v1
      kind: HorizontalPodAutoscaler
      name: nginx-deployment-hpa
   jobs:
   - name: "scale-down"
     schedule: "* 45 11 12 2 *"
     targetSize: 1
```
scenes:
at 11:45 pm

hpa(min/max) | cronhpa | deployment | result
-- | -- | -- | --
5/10 | 1 | 1 | hpa(5/10) deployment 1
----------
```
apiVersion: autoscaling.alibabacloud.com/v1beta1
kind: CronHorizontalPodAutoscaler
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: cronhpa-sample
spec:
   scaleTargetRef:
      apiVersion: autoscaling/v1
      kind: HorizontalPodAutoscaler
      name: nginx-deployment-hpa
   jobs:
   - name: "scale-down"
     schedule: "* * * * * 3"
     targetSize: 1
```
scenes:

hpa(min/max) | cronhpa | deployment | result
-- | -- | -- | --
5/10 | 1 | 1 | hpa(5/10) deployment 1

File: pkg/controller/cronhorizontalpodautoscaler/cronjob.go (hunk 2 above)
Comment (yancl) [reply pkg/controller/cronhorizontalpodautoscaler/cronjob.go]: Great, thanks:)

**Pull request merged**